geomesa-utils/src/main/scala/org/locationtech/geomesa/utils/stats/BinnedArray.scala

```diff
@@ -254,47 +262,45 @@ class BinnedDoubleArray(length: Int, bounds: (jDouble, jDouble)) extends BinnedA
 class BinnedStringArray(length: Int, bounds: (String, String)) extends BinnedArray[String](length, bounds) {
 
   private val (start, end) = {
-    val lower = bounds._1.toLowerCase(Locale.US).replaceAll("[^0-9a-z]", "0")
-    val upper = bounds._2.toLowerCase(Locale.US).replaceAll("[^0-9a-z]", "0")
+    val lower = normalize(bounds._1)
+    val upper = normalize(bounds._2)
     if (lower.length == upper.length) {
       (lower, upper)
     } else if (lower.length < upper.length) {
       (lower.padTo(upper.length, '0'), upper)
     } else {
-      (lower, upper.padTo(upper.length, 'z'))
+      (lower, upper.padTo(lower.length, 'z'))
     }
   }
 
-  require(start < end, s"Upper bound must be greater than lower bound: lower=${bounds._1} upper=${bounds._2}")
-
-  private val firstDiff = start.zip(end).indexWhere { case (l, r) => l != r }
-  private val zeroInt = '0'.toInt
-  private val nineInt = '9'.toInt - zeroInt
-  private val alphaOffset = 'a'.toInt - 10
+  require(start <= end, s"Upper bound must be greater than lower bound: lower=${bounds._1} upper=${bounds._2}")
 
-  def stringToLong(s: String): Long = {
-    val base36 = s.slice(firstDiff, start.length).replaceAll("[^0-9a-z]", "0")
-    if (base36.length > 12) {
-      // won't fit in a long...
-      jLong.parseLong(base36.substring(0, 12), 36)
-    } else {
-      jLong.parseLong(base36, 36)
-    }
+  // # of base 36 numbers we can consider
+  private val sigDigits = math.max(1, math.round(math.log(length) / math.log(36)).toInt)
+  private val prefixLength = if (start == end) start.length else start.zip(end).indexWhere { case (l, r) => l != r }
 
-  }
+  private val minLong = stringToLong(start)
+  private val maxLong = stringToLong(end, 'z')
 
-  def longToString(l: Long): String =
-    start.substring(0, firstDiff) + jLong.toString(l, 36)
+  private val binSize = (maxLong - minLong).toDouble / length
 
-  private val min = stringToLong(start)
-  private val max = stringToLong(end)
+  private def normalize(s: String) = s.toLowerCase(Locale.US).replaceAll("[^0-9a-z]", "0")
 
-  private val binSize = math.max(1, (max - min).toFloat / length)
+  def stringToLong(s: String, padding: Char = '0'): Long = {
+    val normalized = if (s.length >= prefixLength + sigDigits) {
+      s.substring(prefixLength, prefixLength + sigDigits)
+    } else {
+      s.substring(prefixLength).padTo(sigDigits, padding)
+    }
+    jLong.parseLong(normalized, 36)
+  }
+  def longToString(l: Long, padding: Char = '0'): String =
```

Comment:
[picky]  It's probably true that nobody else cares, but using ```l``` in a context where it can easily be mistaken for a ```1``` is potentially confusing.